pydantic/_internal/_typing_extra.py

```diff
@@ -265,24 +265,34 @@ def eval_type_backport(
                 value, globalns, localns
             )
     except TypeError as e:
-        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
-            raise
-        try:
-            from eval_type_backport import eval_type_backport
-        except ImportError:
-            raise TypeError(
-                f'You have a type annotation {value.__forward_arg__!r} '
-                f'which makes use of newer typing features than are supported in your version of Python. '
-                f'To handle this error, you should either remove the use of new syntax '
-                f'or install the `eval_type_backport` package.'
-            ) from e
+        if isinstance(value, typing.ForwardRef):
+            if is_backport_fixable_error(e):
+                try:
+                    from eval_type_backport import eval_type_backport
+                except ImportError:
+                    raise TypeError(
+                        f'Unable to evaluate type annotation {value.__forward_arg__!r}. If you are making use '
+                        'of the new typing syntax (unions using `|` since Python 3.10 or builtins subscripting '
+                        'since Python 3.9), you should either replace the use of new syntax with the existing '
+                        '`typing` constructs or install the `eval_type_backport` package.'
+                    ) from e
+
+                return eval_type_backport(value, globalns, localns, try_default=False)
```

Comment:
On Python 3.8, we will still have the unclear `TypeError: 'type' object is not subscriptable`. We could add yet another try/expect block but this starts to really hurt readability here.

Reply:
I'll try to figure out a better way

Reply:
```python
def eval_type_with_helpful_errors(
    value: Any,
    globalns: dict[str, Any] | None = None,
    localns: dict[str, Any] | None = None,
    type_params: tuple[Any] | None = None,
) -> Any:
    try:
        return _eval_type_with_backport_package(value, globalns, localns, type_params)
    except Exception as e:
        message = f'Unable to evaluate type annotation {value.__forward_arg__!r}.'
        if sys.version_info >= (3, 11):
            e.add_note(message)
        else:
            raise TypeError(message) from e


def _eval_type_with_backport_package(
    value: Any,
    globalns: dict[str, Any] | None = None,
    localns: dict[str, Any] | None = None,
    type_params: tuple[Any] | None = None,
) -> Any:
    try:
        return _eval_type_with_type_params(value, globalns, localns, type_params)
    except TypeError as e:
        if not (isinstance(value, typing.ForwardRef) and is_backport_fixable_error(e)):
            raise

        try:
            from eval_type_backport import eval_type_backport
        except ImportError:
            raise TypeError(
                f'Unable to evaluate type annotation {value.__forward_arg__!r}. If you are making use '
                'of the new typing syntax (unions using `|` since Python 3.10 or builtins subscripting '
                'since Python 3.9), you should either replace the use of new syntax with the existing '
                '`typing` constructs or install the `eval_type_backport` package.'
            ) from e

        return eval_type_backport(value, globalns, localns, try_default=False)


def _eval_type_with_type_params(
    value: Any,
    globalns: dict[str, Any] | None = None,
    localns: dict[str, Any] | None = None,
    type_params: tuple[Any] | None = None,
) -> Any:
    if sys.version_info >= (3, 13):
        return typing._eval_type(  # type: ignore
            value, globalns, localns, type_params=type_params
        )
    else:
        return typing._eval_type(  # type: ignore
            value, globalns, localns
        )
```

Reply:
@Viicos I suggested `except Exception` in `eval_type_with_helpful_errors` so that there would be helpful notes for other errors. We're seeing a mysterious `NameError: name 'Dict' is not defined` and we don't know where it's coming from. The final code in the PR is still only catching `TypeError` so no note is added.

Reply:
The thing is `eval_type_backport` is used by `eval_type_lenient` (defined above), and is the one responsible for catching (and not erroring) `NameError`'s

Reply:
But `eval_type_lenient` is not the only place that calls `eval_type_backport`, so NameError can still bubble up.

`raise type(e)(message) from e` should work.